File: returned/python/today5/set_matrix_zeroes.py
# https://leetcode.com/problems/set-matrix-zeroes/description/
from typing import List

class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        row_count = len(matrix)
        col_count = len(matrix[0])

        for i in range(row_count):
            for j in range(col_count):
                if matrix[i][j] == 0:
                    matrix[i][j] = 'T'
        
        for i in range(row_count):
            for j in range(col_count):
                if matrix[i][j] == 'T':
                    for r in range(row_count):
                        if matrix[r][j] != 'T':
                            matrix[r][j] = 0
                    for c in range(col_count):
                        if matrix[i][c] != 'T':
                            matrix[i][c] = 0
        
        for i in range(row_count):
            for j in range(col_count):
                if matrix[i][j] == 'T':
                    matrix[i][j] = 0

File: typescript/solved/set-matrix-zeroes.ts
/**
 * https://leetcode.com/problems/set-matrix-zeroes/description/
 */
export { setZeroes };

function setZeroes(matrix: number[][]): void {
  const rows = matrix.length;
  const cols = matrix[0].length;
  const zeros: ZeroCell[] = [];

  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      if (matrix[row][col] === 0) {
        zeros.push({ row, col });
      }
    }
  }

  for (let z = 0; z < zeros.length; z++) {
    setZero(matrix, zeros[z]);
  }
}

function setZero(matrix: number[][], zeroCell: ZeroCell): void {
  const rows = matrix.length;
  const cols = matrix[0].length;

  for (let col = 0; col < cols; col++) {
    matrix[zeroCell.row][col] = 0;
  }

  for (let row = 0; row < rows; row++) {
    matrix[row][zeroCell.col] = 0;
  }
}

interface ZeroCell {
  row: number;
  col: number;
}
